src/main/scala/org/opennetworkinsight/netflow/FlowPostLDA.scala

```diff
@@ -63,67 +65,51 @@ object FlowPostLDA {
-    val dataScored = minimumScore(dataWithDestScore)
-
-    logger.info("Persisting data")
-    val filteredDF = dataScored.filter(MinimumScore + " <" + threshold)
-
-    val count = filteredDF.count
+    val docToTopicMixRDD: RDD[(String, Array[Double])] = sc.parallelize(docToTopicMix.toSeq)
 
-    val takeCount  = if (topK == -1 || count < topK) {
-      count.toInt
-    } else {
-      topK
-    }
-
-    val minimumScoreIndex = filteredDF.schema.fieldNames.indexOf(MinimumScore)
-
-    class DataOrdering() extends Ordering[Row] {
-      def compare(row1: Row, row2: Row) = row1.getDouble(minimumScoreIndex).compare(row2.getDouble(minimumScoreIndex))
-    }
+    val docToTopicMixDF = docToTopicMixRDD.map({ case (doc, probabilities) => DocTopicMix(doc, probabilities) }).toDF
 
-    implicit val ordering = new DataOrdering()
-
-    val top: Array[Row] = filteredDF.rdd.takeOrdered(takeCount)
+    val words = sc.broadcast(wordToProbPerTopic)
 
-    val outputRDD = sc.parallelize(top).sortBy(row => row.getDouble(minimumScoreIndex))
+    val dataWithSrcScore = score(sc, dataWithWord, docToTopicMixDF, words, SourceScore, SourceIP, SourceProbabilities, SourceWord)
+    val dataWithDestScore = score(sc, dataWithSrcScore, docToTopicMixDF, words, DestinationScore, DestinationIP, DestinationProbabilities, DestinationWord)
+    val dataScored = minimumScore(dataWithDestScore)
 
-    // Using dropRight as we don't need last column MinimumScore, only SourceScore and DestinationScore
-    outputRDD.map(row => Row.fromSeq(row.toSeq.dropRight(1))).map(_.mkString(outputDelimiter)).saveAsTextFile(resultsFilePath)
+    logger.info("Persisting data")
+    val filteredDF = dataScored.filter(MinimumScore + " < " + threshold)
+    filteredDF.orderBy(MinimumScore).limit(topK).rdd.map(row => Row.fromSeq(row.toSeq.dropRight(1))).map(_.mkString(outputDelimiter)).saveAsTextFile(resultsFilePath)
 
     logger.info("Flow post LDA completed")
-
   }
 
   def score(sc: SparkContext,
             dataFrame: DataFrame,
-            ipToTopicMixes: Map[String, Array[Double]],
-            wordToProbPerTopic: Map[String, Array[Double]],
+            docToTopicMixesDF: DataFrame,
+            words: Broadcast[Map[String, Array[Double]]],
             newColumnName: String,
             ipColumnName: String,
+            ipProbabilitiesColumnName: String,
             wordColumnName: String): DataFrame = {
 
-    val topics = sc.broadcast(ipToTopicMixes)
-    val words = sc.broadcast(wordToProbPerTopic)
+    val dataWithIpProbJoin = dataFrame.join(docToTopicMixesDF, dataFrame(ipColumnName) === docToTopicMixesDF(Doc))
 
-    def scoreFunction(ip: String, word: String): Double = {
-      val uniformProb = Array.fill(20) {
-        0.05d
-      }
+    var newSchemaColumns = dataFrame.schema.fieldNames :+ Probabilities + " as " + ipProbabilitiesColumnName
+    val dataWithIpProb = dataWithIpProbJoin.selectExpr(newSchemaColumns: _*)
```

Comment:
I can't find it quickly in the spark dataframe documentation, and it looks arcane to me... maybe it's standard to SQL people